internal/receiver/smartagentreceiver/output.go

```diff
@@ -34,67 +35,117 @@ import (
-	converter             Converter
-	extraDimensions       map[string]string
+	receiverName         string
+	nextConsumer         consumer.MetricsConsumer
+	nextDimensionClients []*metadata.MetadataExporter
+	nextEventClients     []*consumer.LogsConsumer
+	logger               *zap.Logger
+	converter            Converter
+	extraDimensions      map[string]string
 }
 
 var _ types.Output = (*Output)(nil)
 var _ types.FilteringOutput = (*Output)(nil)
 
 func NewOutput(config Config, nextConsumer consumer.MetricsConsumer, host component.Host, logger *zap.Logger) *Output {
-	metadataExporters := getMetadataExporters(config, host, nextConsumer, logger)
+	metadataExporters := getMetadataExporters(config, host, &nextConsumer, logger)
+	logConsumers := getLogsConsumers(config, host, &nextConsumer, logger)
 	return &Output{
-		receiverName:          config.Name(),
-		nextConsumer:          nextConsumer,
-		nextMetadataConsumers: metadataExporters,
-		logger:                logger,
-		converter:             Converter{logger: logger},
-		extraDimensions:       map[string]string{},
+		receiverName:         config.Name(),
+		nextConsumer:         nextConsumer,
+		nextDimensionClients: metadataExporters,
+		nextEventClients:     logConsumers,
+		logger:               logger,
+		converter:            Converter{logger: logger},
+		extraDimensions:      map[string]string{},
 	}
 }
 
-func getMetadataExporters(config Config, host component.Host, nextConsumer consumer.MetricsConsumer, logger *zap.Logger) []*metadata.MetadataExporter {
+// getMetadataExporters walks through obtained Config.MetadataClients and returns all matching registered MetadataExporters,
+// if any.  At this time the SignalFx exporter is the only supported use case and adopter of this type.
+func getMetadataExporters(
+	config Config, host component.Host, nextConsumer *consumer.MetricsConsumer, logger *zap.Logger,
+) []*metadata.MetadataExporter {
 	var exporters []*metadata.MetadataExporter
 
-	if config.MetadataClients != nil {
-		builtExporters := host.GetExporters()[configmodels.MetricsDataType]
-		for _, client := range *config.MetadataClients {
-			var exporter component.Exporter
-			for k, v := range builtExporters {
-				if k.Name() == client {
-					exporter = v
-					break
-				}
-			}
-			if metadataExporter, ok := exporter.(metadata.MetadataExporter); ok {
-				exporters = append(exporters, &metadataExporter)
-			} else {
-				logger.Warn(
-					"provided metadataClients item not a MetadataExporter.  cannot send dimension updates",
-					zap.String("client", client),
-				)
-			}
+	metadataExporters := getClientsFromMetricsExporters(config.DimensionClients, host, nextConsumer, "dimensionClients", logger)
+	for _, client := range metadataExporters {
+		if metadataExporter, ok := (*client).(metadata.MetadataExporter); ok {
+			exporters = append(exporters, &metadataExporter)
+		} else {
+			logger.Info("cannot send dimension updates to dimension client", zap.Any("client", *client))
```

Comment:
Assuming that metadata clients are only configured for receiver instances that wrap monitors emitting metadata, should this be logged as a warning? Same with the log for event clients.

Reply:
This will occur for all SA receivers since they all have an Output and their nextConsumer default check will be attempted if no clients are configured.